x-pack/plugins/security_solution/public/detection_engine/rule_management/components/rule_details/json_diff/json_diff.test.tsx

```diff
@@ -0,0 +1,331 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React from 'react';
+import { EuiThemeProvider } from '@elastic/eui';
+import { render, screen } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+import { uniq, sortBy, isEqual } from 'lodash';
+
+import { RuleDiffTab } from '../rule_diff_tab';
+import { savedRuleMock } from '../../../logic/mock';
+import type { RuleResponse } from '../../../../../../common/api/detection_engine/model/rule_schema/rule_schemas.gen';
+import { COLORS } from './constants';
+
+describe('Rule upgrade workflow: viewing rule changes in JSON diff view', () => {
+  it.each(['light', 'dark'] as const)(
+    'User can see precisely how property values would change after upgrade - %s theme',
+    (colorMode) => {
+      const oldRule: RuleResponse = {
+        ...savedRuleMock,
+      };
+
+      const newRule: RuleResponse = {
+        ...savedRuleMock,
+      };
+
+      /* Changes to test line update */
+      oldRule.version = 1;
+      newRule.version = 2;
+
+      /* Changes to test line removal */
+      oldRule.author = ['Alice', 'Bob', 'Charlie'];
+      newRule.author = ['Alice', 'Charlie'];
+
+      /* Changes to test line addition */
+      delete oldRule.license;
+      newRule.license = 'GPLv3';
+
+      const ThemeWrapper: React.FC<{}> = ({ children }) => (
+        <EuiThemeProvider colorMode={colorMode}>{children}</EuiThemeProvider>
+      );
+
+      const { container } = render(<RuleDiffTab oldRule={oldRule} newRule={newRule} />, {
+        wrapper: ThemeWrapper,
+      });
+
+      /*
+      Helper function to find an element with a particular text content.
+      React Testing Library's doesn't provide an easy way to search by text if text is split into multiple DOM elements.
+    */
+      const getChildByTextContent = (parent: Element, textContent: string): HTMLElement => {
+        return Array.from(parent.querySelectorAll('*')).find(
+          (childElement) => childElement.textContent === textContent
+        ) as HTMLElement;
+      };
```

Comment:
Any reason this function and `findDiffLineContaining` cannot be extracted out to the top as helper functions, outside of the tests? This one is redefined on each `it` clause.

Reply:
I thought it would be a bit easier to read if these functions are local to where they are used. But I understand that it's a matter of preference and the benefit is small, so I have no trouble moving these to the top level.